File: BAEKJOON/java/15659_연산자_끼워넣기_(3).java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class Main {

    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static int n, maxNum, minNum;
    public static int[] nums, ops;
    public static List<Character> operators, list;
    public static boolean[] visit;

    public static void main(String[] args) throws IOException {
        input();
        solution();

        bw.flush();
        bw.close();
    }

    private static void input() throws IOException {
        n = Integer.parseInt(br.readLine());
        nums = new int[n];
        ops = new int[4];

        String[] st = br.readLine().split(" ");

        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(st[i]);
        }

        st = br.readLine().split(" ");

        for (int i = 0; i < 4; i++) {
            ops[i] = Integer.parseInt(st[i]);
        }
    }
    private static void solution() throws IOException {
        operators = makeOps();
        visit = new boolean[operators.size()];
        list = new ArrayList<>();

        minNum = 1000000001;
        maxNum = -1000000001;

        //순열
        dfs(0);

        bw.write(String.valueOf(maxNum) + "\n" + String.valueOf(minNum));
    }

    private static List<Character> makeOps() {
        List<Character> ret = new ArrayList<>();
        char[] op = {'+', '-', '*', '/'};

        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < ops[i]; j++) {
                ret.add(op[i]);
            }
        }

        return ret;
    }

    private static void dfs(int cnt) {
        if (cnt == n - 1) {
            int result = calculate();

            minNum = Math.min(minNum, result);
            maxNum = Math.max(maxNum, result);

            return;
        }

        for (int i = 0; i < operators.size(); i++) {
            if (visit[i]) continue;
            visit[i] = true;
            list.add(operators.get(i));
            dfs(cnt + 1);
            visit[i] = false;
            list.remove(list.size() - 1);
        }
    }

    private static int calculate() {
        List<Integer> calNums = new ArrayList<>();
        List<Character> calOps = new ArrayList<>();

        int idx = 0;
        calNums.add(nums[idx++]);

        //곱하기, 나누기 먼저 계산
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i) != '*' && list.get(i) != '/') {
                calOps.add(list.get(i));
                calNums.add(nums[idx++]);
            } else {
                int cur = calNums.get(calNums.size() - 1);
                calNums.remove(calNums.size() - 1);

                if (list.get(i) == '*') {
                    calNums.add(cur * nums[idx++]);
                } else {
                    calNums.add(cur / nums[idx++]);
                }
            }
        }

        //더하기, 빼기 계산
        idx = 0;
        int result = calNums.get(idx++);

        for (int i = 0; i < calOps.size(); i++) {
            if (calOps.get(i) == '+') {
                result += calNums.get(idx++);
            } else {
                result -= calNums.get(idx++);
            }
        }

        return result;
    }
}

File: BAEKJOON/cpp/15659_연산자_끼워넣기_(3).cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n, maxNum, minNum;
vector<int> nums, ops;
vector<char> operators, v;
vector<bool> visit;

void input() {
	cin >> n;

	int a;
	for (int i = 0; i < n; i++) {
		cin >> a;
		nums.push_back(a);
	}

	for (int i = 0; i < 4; i++) {
		cin >> a;
		ops.push_back(a);
	}
}

vector<char> makeOps() {
	vector<char> ret;
	char op[] = {'+', '-', '*', '/'};

	for (int i = 0; i < 4; i++) {
		for (int j = 0; j < ops[i]; j++) {
			ret.push_back(op[i]);
		}
	}

	return ret;
}

int calculate() {
	vector<int> calNums;
	vector<char> calOps;

	int idx = 0;
	calNums.push_back(nums[idx++]);

	//곱하기, 나누기 먼저 계산
	for (int i = 0; i < v.size(); i++) {
		if (v[i] != '*' && v[i] != '/') {
			calOps.push_back(v[i]);
			calNums.push_back(nums[idx++]);
		}
		else {
			int cur = calNums[calNums.size() - 1];
			calNums.pop_back();

			if (v[i] == '*') {
				calNums.push_back(cur * nums[idx++]);
			}
			else {
				calNums.push_back(cur / nums[idx++]);
			}
		}
	}

	//더하기, 빼기 계산
	idx = 0;
	int result = calNums[idx++];

	for (int i = 0; i < calOps.size(); i++) {
		if (calOps[i] == '+') {
			result += calNums[idx++];
		}
		else {
			result -= calNums[idx++];
		}
	}

	return result;
}

void dfs(int cnt) {
	if (cnt == n - 1) {
		int result = calculate();

		minNum = min(minNum, result);
		maxNum = max(maxNum, result);

		return;
	}

	for (int i = 0; i < operators.size(); i++) {
		if (visit[i] == true) continue;
		visit[i] = true;
		v.push_back(operators[i]);
		dfs(cnt + 1);
		visit[i] = false;
		v.pop_back();
	}
}

void solution() {
	operators = makeOps();
	visit.assign(operators.size(), false);

	minNum = 1000000001;
	maxNum = -1000000001;
	
	//순열
	dfs(0);

	cout << maxNum << "\n" << minNum;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	input();
	solution();

	return 0;
}